lib/compiler/passes/inference-match-result.js

```diff
@@ -0,0 +1,108 @@
+const GrammarError = require("../../grammar-error");
+
+// Inference match result of the rule. Can be:
+// -1: negative result, always fails
+//  0: neutral result, may be fail, may be match
+//  1: positive result, always match
+function inferenceMatchResult(ast) {
+  function sometimesMatch(node) { return (node.match = 0); }
+  function alwaysMatch(node) {
+    inference(node.expression);
+
+    return (node.match = 1);
+  }
+
+  function inferenceExpression(node) {
+    return (node.match = inference(node.expression));
+  }
+  function inferenceElements(elements, forChoice) {
+    const length = elements.length;
+    let always = 0;
+    let never = 0;
+
+    for (let i = 0; i < length; ++i) {
+      const result = inference(elements[i]);
+
+      if (result > 0) { ++always; }
+      if (result < 0) { ++never;  }
+    }
+
+    if (always === length) {
+      return 1;
+    }
+    if (forChoice) {
+      return never === length ? -1 : 0;
+    }
+
+    return never > 0 ? -1 : 0;
+  }
+
+  const inference = visitor.build({
+    rule(node) {
+      let oldResult;
+      let count = 0;
+
+      if (typeof node.match === "undefined") {
+        node.match = 0;
+        do {
+          oldResult = node.match;
+          node.match = inference(node.expression);
+          // 6 == 3! -- permutations count for all transitions from one match
+          // state to another.
+          // After 6 iterations the cycle with guarantee begins
+          if (++count > 6) {
+            throw new GrammarError(
+              "Infinity cycle detected when trying evaluate node match result",
+              node.location
+            );
+          }
+        } while (oldResult !== node.match);
+      }
+
+      return node.match;
+    },
+    named:        inferenceExpression,
+    choice(node) {
+      return (node.match = inferenceElements(node.alternatives, true));
+    },
+    action:       inferenceExpression,
+    sequence(node) {
+      return (node.match = inferenceElements(node.elements, false));
+    },
+    labeled:      inferenceExpression,
+    text:         inferenceExpression,
+    simple_and:   inferenceExpression,
+    simple_not(node) {
+      return (node.match = -inference(node.expression));
+    },
+    optional:     alwaysMatch,
+    zero_or_more: alwaysMatch,
+    one_or_more:  inferenceExpression,
+    group:        inferenceExpression,
+    semantic_and: sometimesMatch,
+    semantic_not: sometimesMatch,
+    rule_ref(node) {
+      const rule = asts.findRule(ast, node.name);
+
+      return (node.match = inference(rule));
+    },
+    literal(node) {
+      // Empty literal always match on any input
+      return (node.match = node.value.length === 0 ? 1 : 0);
+    },
+    class(node) {
+      // Empty character class never match on any input
+      return (node.match = node.parts.length === 0 ? -1 : 0);
+    },
+    // |any| not match on empty input
+    any:          sometimesMatch
+  });
+
+  inference(ast);
+}
```

Comment:
are you attempting to add a nearly 100 line function to the codebase, after telling someone else that their 35 line PR was too complicated for you to follow, and couldn't go in?

Reply:
Your PR was complicated not in the count of lines, but in count of commits, most of them looks like you doing try-and-run procedure. I do not forbid you to do this in your work, while you are looking for a solution or awaiting for a review comments, moreover, sometimes that is impossible without that because of the nature of CI changing moments.

I do the same things internally, but this does not mean that after addressing review comments we should keep all that crap in the main history. It has absolutely no value, hard for understanding (because that is your and only your brainstorming which is relevant only for a concrete time moment) and for maintaining.

That code is just couldn't be written in more than 100 lines because of the project codestyle.

----
In general, this claim of yours looks like a child offense, please do not do so. It only makes it worse for you. No offense.

Reply:
> Your PR was complicated not in the count of lines, but in count of commits

I'm sorry that you find github challenging.  I gave you clear instructions on how to view it merged.  I hope someday you learn this valuable skill.

It's less than a third the size of your single function.  

&nbsp;

> most of them looks like you doing try-and-run procedure.

Each one of them describes a separate step in the process.  None of them roll back the behavior of any of the others.

I'm sorry that you find guessing and insults to be appropriate.  The reason this approach is a github standard is so that people can look at individual steps.

&nbsp;

> I do the same things internally, but this does not mean that after addressing review comments we should keep all that crap in the main history.

I'm sorry that you find Github challenging.  None of that would have gone into the main history, because of "merge on commit," which was already clearly explained to you.

I'm sorry that every time you disagree with something, or don't understand the value that it provides, you misrepresent it as crap, or the result of guesswork, to establish social dominance.

The value is clear, obvious, and quite commonly used.  I hope that some day you can learn from the people trying to teach you how to get more results out of these tools, instead of referring to their work as "crap," "low quality," and "sloppy" in public.

I'd love to see you successfully complete the same work without help, while also avoiding the choices I made that you think are incorrect.  I've been politely inviting you to for several days, and saying that perhaps I have something to learn.

&nbsp;

> It has absolutely no value

You might be surprised.

&nbsp;

> hard for understanding

You can click one tab to make it all go away, friend.  I hope you learn how some day.

&nbsp;

> because that is your and only your brainstorming

It actually isn't.  It's a valuable device for engineers to communicate about individual steps distinctly.

&nbsp;

> and for maintaining.

You keep saying there's a maintenance problem in accepting a PR that will be squashed, if it hasn't been squashed yet.

I'd be delighted to learn what, specifically that is.

Reply:
Amusingly, what you misrepresent as a "try and run approach" is exactly the splitting up you wanted into a bunch of PRs

It's the thing you called for; you just don't realize you can have both at the same time

Reply:
> It's less than a third the size of your single function.

I think it is better to discuss each PR problems in their own tickets.

> I'm sorry that you find guessing and insults to be appropriate. 

I'm sorry if you heard this in my words. I'm not a native speaker, so may be I've choose the wrong words. Forgive me.

> You can click one tab to make it all go away, friend. I hope you learn how some day.
> It actually isn't. It's a valuable device for engineers to communicate about individual steps distinctly.

It would be if each commit was followed by clear and complete description why it is done in that way and why is can not be done in the other. I've not saw that that is why I think it is your internal work, which is not necessary to be in the final history.